File: fp-in-scala/src/main/scala/fpinscala/exercises/state/State.scala
package fpinscala.exercises.state

trait RNG:
  def nextInt: (Int, RNG) // Should generate a random `Int`. We'll later define other functions in terms of `nextInt`.

object RNG:
  // NB - this was called SimpleRNG in the book text

  case class Simple(seed: Long) extends RNG:
    def nextInt: (Int, RNG) =
      val newSeed =
        (seed * 0x5deece66dL + 0xbL) & 0xffffffffffffL // `&` is bitwise AND. We use the current seed to generate a new seed.
      val nextRNG = Simple(
        newSeed
      ) // The next state, which is an `RNG` instance created from the new seed.
      val n =
        (newSeed >>> 16).toInt // `>>>` is right binary shift with zero fill. The value `n` is our new pseudo-random integer.
      (
        n,
        nextRNG
      ) // The return value is a tuple containing both a pseudo-random integer and the next `RNG` state.

  type Rand[+A] = RNG => (A, RNG)

  val int: Rand[Int] = _.nextInt

  def unit[A](a: A): Rand[A] =
    rng => (a, rng)

  def map[A, B](s: Rand[A])(f: A => B): Rand[B] =
    rng =>
      val (a, rng2) = s(rng)
      (f(a), rng2)

  // Exercise 6.1: 非負整数をランダム生成する関数 `nonNegativeInt` を実装せよ。

  def nonNegativeInt(rng: RNG): (Int, RNG) = ???

  // Exercise 6.2: 0以上1未満の浮動小数点数をランダム生成する関数 `double` を実装せよ。

  def double(rng: RNG): (Double, RNG) = ???

  // Exercise 6.3: 整数と浮動小数点数の組をランダム生成する関数 `intDouble` と `doubleInt` を実装せよ。
  // また、浮動小数点数の3つ組をランダム生成する関数 `double3` を実装せよ。

  def intDouble(rng: RNG): ((Int, Double), RNG) = ???

  def doubleInt(rng: RNG): ((Double, Int), RNG) = ???

  def double3(rng: RNG): ((Double, Double, Double), RNG) = ???

  // Exercise 6.4: 引数で指定された要素数の整数リストをランダム生成する関数 `ints` を実装せよ。

  def ints(count: Int)(rng: RNG): (List[Int], RNG) = ???

  // Exercise 6.5: `map` を用いて `double` を実装せよ。

  // Exercise 6.6: 関数 `map2` を実装せよ。

  def map2[A, B, C](ra: Rand[A], rb: Rand[B])(f: (A, B) => C): Rand[C] = ???

  // Exercise 6.7: 関数 `sequence` を実装せよ。

  def sequence[A](rs: List[Rand[A]]): Rand[List[A]] = ???

  // Exercise 6.8: 関数 `flatMap` を実装せよ。

  def flatMap[A, B](r: Rand[A])(f: A => Rand[B]): Rand[B] = ???

  // Exercise 6.9: `flatMap` を用いて `map`, `map2` を実装せよ。

  def mapViaFlatMap[A, B](r: Rand[A])(f: A => B): Rand[B] = ???

  def map2ViaFlatMap[A, B, C](ra: Rand[A], rb: Rand[B])(
      f: (A, B) => C
  ): Rand[C] = ???

opaque type State[S, +A] = S => (A, S)

object State:
  extension [S, A](underlying: State[S, A])
    def run(s: S): (A, S) = underlying(s)

    // Exercise 6.10: 拡張メソッド `map`, `map2`, `flatMap` を実装せよ。
    // また、関数 `unit`, `sequence`, `traverse` を実装せよ。

    def map[B](f: A => B): State[S, B] =
      ???

    def map2[B, C](sb: State[S, B])(f: (A, B) => C): State[S, C] =
      ???

    def flatMap[B](f: A => State[S, B]): State[S, B] =
      ???

  def apply[S, A](f: S => (A, S)): State[S, A] = f

  def modify[S](f: S => S): State[S, Unit] =
    for
      s <- get // Gets the current state and assigns it to `s`.
      _ <- set(f(s)) // Sets the new state to `f` applied to `s`.
    yield ()

  def get[S]: State[S, S] = s => (s, s)

  def set[S](s: S): State[S, Unit] = _ => ((), s)

// Exercise 6.11: Stateを用いて以下のルールを満たすキャンディ販売機の振る舞いをシミュレートする関数 `simulateMachine` を実装せよ。
// `simulateMachine` は入力リストを受け取って販売機の最終的なキャンディの個数とコインの枚数のペアを返す。
// ルール:
//   - 販売機がロックされている(`locked = true`)とき、ノブを回し(Input.Turn)ても販売機は反応しない
//   - 販売機がロックされている(`locked = true`)とき、コインを投入する(Input.Coin)と販売機のロックが外れてコインが1枚増える
//   - 販売機がロックされていない(`locked = false`)とき、ノブを回す(Input.Turn)と販売機のロックが掛かってキャンディが1個減る
//   - 販売機がロックされていない(`locked = false`)とき、コインを投入し(Input.Coin)ても販売機は反応しない
//   - 販売機にキャンディが残っていない(`candies = 0`)とき、コインを投入し(Input.Coin)てもノブを回し(Input.Turn)ても販売機は反応しない

enum Input:
  case Coin, Turn

case class Machine(locked: Boolean, candies: Int, coins: Int)

object Candy:
  def simulateMachine(inputs: List[Input]): State[Machine, (Int, Int)] = ???

File: fp-in-haskell/src/FpInHaskell/Answers/State/State.hs
module FpInHaskell.Answers.State.State (
    State (..),
    unit,
    map,
    map2,
    flatMap,
    sequence,
    traverse,
    get,
    set,
    modify,
) where

import Prelude hiding (map, sequence, traverse)

-- Prelude には同名の `map`/`sequence`/`traverse` がすでにあるため、上の
-- `import Prelude hiding (...)` でそれらを読み込み対象から外し、ここでは `State` に対する
-- 同名の演習関数を自前で定義している。

-- 状態 `s` を持ち回しながら値 `a` を計算する計算を表す型。`newtype` を使うことで、
-- 実行時のコストなしに元の関数型 `s -> (a, s)` と見分けられる型を作れる。
-- フィールド名 `runState` で元の関数を取り出して実行できる。
newtype State s a = State {runState :: s -> (a, s)}

unit :: a -> State s a
unit x = State (\st -> (x, st))

-- Exercise 6.10: `map`, `map2`, `flatMap` を実装せよ。また、関数 `unit`, `sequence`, `traverse` を実装せよ。
--
-- Prelude の `map` と同じ引数順(関数、State の順)。
map :: (a -> b) -> State s a -> State s b
map f state = flatMap (unit . f) state

map2 :: (a -> b -> c) -> State s a -> State s b -> State s c
map2 f sa sb = flatMap (\x -> map (f x) sb) sa

flatMap :: (a -> State s b) -> State s a -> State s b
flatMap f (State run) = State (\st0 -> let (x, st1) = run st0 in runState (f x) st1)

sequence :: [State s a] -> State s [a]
sequence = foldr (\st acc -> map2 (:) st acc) (unit [])

traverse :: (a -> State s b) -> [a] -> State s [b]
traverse f = foldr (\x acc -> map2 (:) (f x) acc) (unit [])

-- `map`/`flatMap` を `Functor`/`Applicative`/`Monad` に接続しておく。こうすると `modify`
-- (下で定義)や次のファイルの `Candy.simulateMachine` を `do` 記法で書ける。`do` 記法は
-- `>>=`(bind、ここでは上で実装した `flatMap` と同じもの)へのシンタックスシュガーだ。
instance Functor (State s) where
    fmap = map

instance Applicative (State s) where
    pure = unit

    -- ここでの `$` は括弧を避けるための記法ではなく、「関数を引数に適用する」という演算そのものを
    -- 値として `map2` に渡している。`map2 (\f x -> f x) sf sa` と同じ意味になる。
    sf <*> sa = map2 ($) sf sa

instance Monad (State s) where
    st >>= f = flatMap f st

-- 現在の状態を値として取り出す。
get :: State s s
get = State (\st -> (st, st))

-- 状態を `newState` に置き換える。
set :: s -> State s ()
set newState = State (const ((), newState))

-- 関数 `f` で状態を更新する。`do` 記法の内部では `get`/`set` の呼び出しが `flatMap` で
-- 繋がれている。
modify :: (s -> s) -> State s ()
modify f = do
    st <- get
    set (f st)
